nntrainer/optimizers/optimizer_devel.cpp

```diff
@@ -1,51 +1,27 @@
+// SPDX-License-Identifier: Apache-2.0
 /**
- * Copyright (C) 2020 Samsung Electronics Co., Ltd. All Rights Reserved.
+ * Copyright (C) 2020 Parichay Kapoor <pk.kapoor@samsung.com>
  *
- * Licensed under the Apache License, Version 2.0 (the "License");
- * you may not use this file except in compliance with the License.
- * You may obtain a copy of the License at
- *   http://www.apache.org/licenses/LICENSE-2.0
- * Unless required by applicable law or agreed to in writing, software
- * distributed under the License is distributed on an "AS IS" BASIS,
- * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
- * See the License for the specific language governing permissions and
- * limitations under the License.
- *
- *
- * @file	optimizer.cpp
+ * @file	optimizer_devel.cpp
```

Comment:
I wonder if we are keeping `optimizer_devel.cpp` ? shouldn't this move to `optimizer_impl.cpp`?

Reply:
`optimizer_devel` provides the essential interface which the user can extend. Further, all the other components of nntrainer must use `optimizer_devel` to interact with the optimizer.
`optimizer_impl` is a base implementation of the optimizer which other optimizers, like adam and sgd can extend. The public interface of `optimizer_impl` can have extra functions to support the implementation of optimizers or their unittests.

A user who wants to build a custom optimizer must support the interface of `nntrainer_devel` but is free to extend `nntrainer_devel` or `nntrainer_impl`.

Reply:
Yes, I mean if it starts to have any members or implementation, wouldn't it be hard to extend?
I think interface by definition should have only pure virtuals, and that's a reason to make optimizer_impl.h

Reply:
eg) If you are creating a mock that returns 1 for `getLearningRate()`

Reply:
`learning_rate` can be moved to `optimizer_impl`, but then every custom optimizer will have to add a learning rate and its getters. If you think it's better that way, it can be done. It is added in `optimizer_devel` as every optimizer by definition must have a learning rate.

Reply:
I think the `optimizer_devel` should remain pure virtual for all other methods too.
If we have implementation for the `optimizer_devel`, the developer will need to know about the implementation of `optimizer_devel`. Making abstraction (`optimizer_devel.h`) to depend on the concrete (`optimizer_devel.cpp`)

Some implementations are very common so we have a pre-implemented abstract class which would be `optimizer_impl`. But if we have to extend the interface, there shouldn't be any implementations ready that is depending upon.

Reply:
> If we have implementation for the optimizer_devel, the developer will need to know about the implementation of optimizer_devel.

Assuming all the methods are virtual, the developer does not need to know any implementation of the `optimizer_devel` (I think some functions are not currently virtual and will be done). If the default implementation is removed, the developer still needs to know the necessary semantics of the optimizer - for example, `applyGradients()` is responsible to add the gradient corresponding to the regularization. Without any default implementation, such information will have to be provided in the comments of the interface anyway so that the user can add that code. 

The current default implementation adds a state (which is the learning rate and can be removed). But I can't see any disadvantage with default interface implementation (assume to be stateless although having some state which every extension of that interface will have should be fine too).

Reply:
We made an offline discussion to remove state and make all method virtual but leave default method.